### PR TITLE
[SPARK-9893] User guide with Java test suite for VectorSlicer

### PR DESCRIPTION
**Pull request by yinxusen**
Add user guide for `VectorSlicer`, with Java test suite and Python version VectorSlicer.

Note that Python version does not support selecting by names now.

### CODE REVIEW

**Comment by feynmanliang** [comment]
@yinxusen Sorry, I think there's some merge conflicts. Do you mind rebasing master?

**Comment by feynmanliang** [comment]
Could you also modify the title to include java tests and python API?

File: docs/ml-features.md [hunk 1]
```diff
@@ -1389,3 +1389,145 @@ print(output.select("features", "clicked").first())
...
+
+`VectorSlicer` is a transformer that takes a feature vector and outputs a new feature vector with a sub-array of the original features. It is useful for extracting features from a vector column.
+
+`VectorSlicer` accepts a vector column with a specified indices, then outputs a new vector column whose values are selected via those indices. There are two types of indices, 
+
+ 1. Integer indices that represents the real indices in the vector, `setIndices()`;
+
+ 2. String indices that represents the names of features in the vector, `setNames()`.
+
+Specify by integer and string are both acceptable, moreover, you can use integer index and string name simultaneously. At least one feature must be selected. Duplicate features are not allowed, so there can be no overlap between selected indices and names. Note that if names of features are selected, an exception will be threw out when encountering with empty input attributes.
```
Comment (feynmanliang) [inline docs/ml-features.md]: I know other parts of the user guides don't do this, but we should try to keep lines to 100 characters so future diffs are smaller

File: docs/ml-features.md [hunk 2]
```diff
@@ -1389,3 +1389,145 @@ print(output.select("features", "clicked").first())
...
+~~~
+ userFeatures     | features
+------------------|-----------------------------
+ [0.0, 10.0, 0.5] | [10.0, 0.5]
+ ["f1", "f2", "f3"] | ["f2", "f3"]
+~~~
+
+**NOTE**
+
+`VectorSlicer` of Python version does not supprt selecting by names currently.
```
Comment (feynmanliang) [inline docs/ml-features.md]: nit: "The Python version of `VectorSlicer`..."
supprt -> support

File: docs/ml-features.md [hunk 3]
```diff
@@ -1389,3 +1389,145 @@ print(output.select("features", "clicked").first())
 
 # Feature Selectors
 
+## VectorSlicer
+
+`VectorSlicer` is a transformer that takes a feature vector and outputs a new feature vector with a sub-array of the original features. It is useful for extracting features from a vector column.
+
+`VectorSlicer` accepts a vector column with a specified indices, then outputs a new vector column whose values are selected via those indices. There are two types of indices, 
+
+ 1. Integer indices that represents the real indices in the vector, `setIndices()`;
```
Comment (feynmanliang) [inline docs/ml-features.md]: I would remove the word "real" (i.e. "...that represent the indices into the vector") since it could be confused for real numbers (i.e. real-valued indices, which don't really make sense)

File: docs/ml-features.md (hunk 1 above)
Comment (feynmanliang) [inline docs/ml-features.md]: nit: ***Specification*** by integer and string are both acceptable***. M***oreover,

File: docs/ml-features.md [hunk 4]
```diff
@@ -1389,3 +1389,145 @@ print(output.select("features", "clicked").first())
...
+
+ 2. String indices that represents the names of features in the vector, `setNames()`.
```
Comment (feynmanliang) [inline docs/ml-features.md]: Document that this requires the vector column to have an `AttributeGroup` since the implementation matches on the `name` field of an `Attribute`

**Comment by feynmanliang** [comment]
@yinxusen The documentation looks good. Only things left:

 * Can you update the title of the PR to include python api and java tests?
 * Do you mind resolving merge conflicts?

Thanks!

**Comment by yinxusen** [comment]
@feynmanliang Conflicts fixed.

**Comment by mengxr** [comment]
@yinxusen Could you split this PR into two? One for the user guide with example code and Java test suite, which can be merged into 1.5, and one implements the Python API. We count Python API as new features, and try not to merge them during QA.

File: python/pyspark/ml/feature.py [hunk 5]
```diff
@@ -30,7 +30,83 @@
...
+    # a placeholder to make it appear in the generated doc
+    indices = Param(Params._dummy(), "indices",
+                    "An array of indices to select features from a vector column." +
+                    " There can be no overlap with names.")
+    names = Param(Params._dummy(), "names",
+                  "An array of feature names to select features from a vector column." +
+                  " There can be no overlap with indices.")
+
+    @keyword_only
+    def __init__(self, indices=[], names=[], inputCol=None, outputCol=None):
```
Comment (mengxr) [inline python/pyspark/ml/feature.py]: Do not use `[]` or `{}` as the default value in Python API, because they are mutable and shared among instances. Please use `None` instead. Again, we should move Python API to a separate PR.

**Comment by feynmanliang** [comment]
jenkins test this please

**Comment by feynmanliang** [comment]
LGTM pending tests

**Comment by mengxr** [comment]
Merged into master and branch-1.5. Thanks!
